mixer/pkg/protobuf/yaml/dynamic/auth_test.go

```diff
@@ -0,0 +1,298 @@
+					},
+				},
+			},
+			errorMessage: "cannot get grpc per rpc credentials unexpected tls token source type",
+		},
+		{
+			name:        "tls oauth token",
+			mode:        TLS,
+			headerKey:   "authorization",
+			headerToken: "Bearer test-token",
+			authCfg: &policypb.Authentication{
+				AuthType: &policypb.Authentication_Tls{
+					Tls: &policypb.Tls{
+						CaCertificates: "../testdata/auth/ca.pem",
+						TokenSource: &policypb.Tls_Oauth{
+							Oauth: &policypb.OAuth{
+								ClientId:     "clientid",
+								ClientSecret: tokenPath,
+								TokenUrl:     ts.URL,
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:        "tls oauth token no secret",
+			mode:        TLS,
+			headerKey:   "authorization",
+			headerToken: "Bearer test-token",
+			authCfg: &policypb.Authentication{
+				AuthType: &policypb.Authentication_Tls{
+					Tls: &policypb.Tls{
+						CaCertificates: "../testdata/auth/ca.pem",
+						TokenSource: &policypb.Tls_Oauth{
+							Oauth: &policypb.OAuth{
+								ClientId: "clientid",
+								TokenUrl: ts.URL,
+							},
+						},
+					},
+				},
+			},
+			errorMessage: "cannot get grpc per rpc credentials oauth secret cannot be empty",
+		},
+	}
+
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			var s spy.Server
+			var err error
+			args := getServerArgs(tt.mode, tt.headerKey, tt.headerToken)
+			if s, err = spy.NewNoSessionServer(args); err != nil {
+				t.Fatalf("unable to start Spy %v", err)
+			}
+			s.Run()
+			defer func() {
+				_ = s.Close()
+			}()
+			t.Logf("Started server at: %v", s.Addr())
+
+			adapterConfig := &types.Any{
+				TypeUrl: "@abc",
+				Value:   []byte("abcd"),
+			}
+			metricDi := loadInstance(t, "metric", "template/metric/template_handler_service.descriptor_set",
+				v1beta1.TEMPLATE_VARIETY_REPORT)
+			h, err := BuildHandler("spy",
+				&attributeV1beta1.Connection{
+					Address:        s.Addr().String(),
+					Authentication: tt.authCfg,
+				},
+				false, adapterConfig,
+				[]*TemplateConfig{metricDi}, true)
+			if err != nil {
+				if tt.errorMessage != "" {
+					if !strings.Contains(err.Error(), tt.errorMessage) {
+						t.Errorf("want %v in error message, got %v", tt.errorMessage, err.Error())
+					}
+					return
+				}
+				t.Fatalf("cannot connect to remote handler %v", err)
+			}
+			mi := buildMetricInst(t)
+			if err := h.HandleRemoteReport(context.Background(), []*adapter.EncodedInstance{mi}); err != nil {
+				t.Errorf("get error %v", err)
+			}
+		})
+	}
+}
+
+func getServerArgs(auth AuthMode, headerKey, headerToken string) *spy.Args {
+	args := spy.DefaultArgs()
+	switch auth {
+	case TLS:
+		args.Behavior.RequireTLS = true
+		args.Behavior.HeaderKey = headerKey
+		args.Behavior.HeaderToken = headerToken
+	case MTLS:
+		args.Behavior.RequireMTls = true
+		args.Behavior.InsecureSkipVerification = true
+	}
+	args.Behavior.KeyPath = "../testdata/auth/private.key"
```

Comment:
nit: would be nice to have distinct client and server private certs

Reply:
Done.